### PR TITLE
Not depending on return value of super(). Fixes #9020.

### LINKED ISSUE #9020: Babel rely on value returned by super() but this is not reliable with Edge
## Bug Report

**Current Behavior**

When using some transpiled code, babel rely on `_this` from  `_this = super()` but this bug on Edge is still present with latest version https://github.com/Microsoft/ChakraCore/issues/4663

**Input Code**
- REPL or Repo link if applicable:

```js
export class App extends React.Component {
    state = {
        fruit: 'banana'
    }

    handleClick = (event, ...rest) => {
        this.setState({fruit: 'apple'})
    }

    render() {
        const {fruit} = this.state
        return (
            <h1 onClick={this.handleClick}>I'm a {fruit}</h1>
        )
    }
}
```

become with preset set on `targets: {esmodules: true}`

```jsx
class App extends react__WEBPACK_IMPORTED_MODULE_0___default.a.Component {
  constructor() {
    var _this;

    _this = super(...arguments);
    this.state = {
      fruit: 'banana'
    };

    this.handleClick = function (event) {
      _this.setState({
        fruit: 'apple'
      });
    };
  }

  render() {
    const fruit = this.state.fruit;
    return react__WEBPACK_IMPORTED_MODULE_0___default.a.createElement("h1", {
      onClick: this.handleClick,
      __source: {
        fileName: _jsxFileName,
        lineNumber: 16
      },
      __self: this
    }, "I'm a ", fruit);
  }

}
```

**Expected behavior/code**

The initial behavior is that `_this` should not be `undefined` but I think that is in the hands of ChakraCore team. Is this something babel can do? There is already this issue which is kinda similar https://github.com/babel/babel/issues/8019 but also this one which emphasis on the fact that `_this` should user the value resturned by `super()` #2279

**Environment**
- Babel version(s): v7.1.5
- Node/npm version: Node 8/npm 6
- OS: build Ubuntu 18.04, this test on Windows 10 / Edge 17
- Monorepo no
- How you are using Babel: loader

**Possible Solution**

Another possible solution would be to not transpile on rest/spread & default argument function when `targets: {esmodules: true}` or `targets: {browsers: 'Edge >= 14'}` because its seems they support it

### PR DESCRIPTION
**Pull request by joeldenning**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #9020 
| Patch: Bug Fix?          | 👍 
| Major: Breaking Change?  | 
| Minor: New Feature?      | 
| Tests Added + Pass?      | 👍 
| Documentation PR Link    | N/A
| Any Dependency Changes?  | No
| License                  | MIT

See #9020 and [this comment](https://github.com/babel/babel/issues/9020#issuecomment-440185493) where @nicolo-ribaudo suggested the change that I implemented here. This is my first contribution to babel and I'm not very familiar with the codebase -- I welcome any feedback. I consider this to be a best effort implementation, since I probably am missing some things or don't fully understand the implications of my code changes.

### CODE REVIEW

File: packages/babel-traverse/src/path/conversion.js [hunk 1]
```diff
@@ -461,8 +461,12 @@ function getThisBinding(thisEnvFn, inConstructor) {
         if (supers.has(child.node)) return;
         supers.add(child.node);
 
-        child.replaceWith(
-          t.assignmentExpression("=", t.identifier(thisBinding), child.node),
+        child.insertAfter(
```
Comment (joeldenning) [inline packages/babel-traverse/src/path/conversion.js]: See [this comment](https://github.com/babel/babel/issues/9020#issuecomment-440185493) for why I made this change

File: packages/babel-plugin-transform-arrow-functions/test/fixtures/arrow-functions/this/output.js [hunk 2]
```diff
@@ -15,9 +15,12 @@ class Foo extends function () {} {
     };
 
     if (true) {
-      console.log(_this2 = super(), foo());
+      var _temp;
+
+      console.log((_temp = super(), _this2 = this, _temp), foo());
```
[the remark this answers is not in the document]
Comment (joeldenning) [reply packages/babel-plugin-transform-arrow-functions/test/fixtures/arrow-functions/this/output.js]: 👍 good catch - @nicolo-ribaudo I just updated it to do what you said. Let me know if I misunderstood.

File: packages/babel-traverse/src/path/conversion.js [hunk 3]
```diff
@@ -461,8 +461,13 @@ function getThisBinding(thisEnvFn, inConstructor) {
         if (supers.has(child.node)) return;
         supers.add(child.node);
 
-        child.replaceWith(
-          t.assignmentExpression("=", t.identifier(thisBinding), child.node),
+        child.replaceWithMultiple(
```
Comment (nicolo-ribaudo) [inline packages/babel-traverse/src/path/conversion.js]: It takes an array.

Comment (joeldenning) [reply packages/babel-traverse/src/path/conversion.js]: @nicolo-ribaudo oops! I just fixed that. I don't fully understand the expected output in some of the tests, so I didn't catch that the second argument was being ignored.

**Comment by joeldenning** [comment]
Yeah I was surprised that there weren't more tests affected, too. Are you the one who will merge this @nicolo-ribaudo? Or is that someone else?

**Comment by joeldenning** [comment]
I think the reason that there aren't more tests affected is that, if i understand the code correctly, this only affects `super()` calls in classes that have arrow functions in the constructor, such as in the case of compiled class properties that use the `fn1 = () => {}` syntax. I'm not 100% sure on that, but that's what I saw in my own testing of it.

**Comment by JSteunou** [comment]
Is there a risk this fix break #2279 ?

**Comment by nicolo-ribaudo** [comment]
It shouldn't; this PR doesn't touch transpilation of `super()`.

**Comment by adrienharnay** [comment]
Hey, any chance this could be released? Thanks 🙂

**Pull request merged**